File: VB/Core/Serilog.Sinks.LogViewVB.Core/Extensions/DataStoreLoggerSinkExtensions.vb
Imports System.Runtime.CompilerServices
Imports LogViewerVB.Core
Imports Serilog.Configuration

Public Module DataStoreLoggerSinkExtensions

    <Extension>
    Public Function DataStoreLoggerSink(loggerConfiguration As LoggerSinkConfiguration,
                                        dataStoreProvider As Func(Of ILogDataStore),
                                        Optional configuration As Action(Of DataStoreLoggerConfiguration) = Nothing,
                                        Optional formatProvider As IFormatProvider = Nothing) As LoggerConfiguration

        Return loggerConfiguration.Sink(New DataStoreLoggerSink(dataStoreProvider, GetConfig(configuration), formatProvider))

    End Function

    Private Function GetConfig(configuration As Action(Of DataStoreLoggerConfiguration)) As Func(Of DataStoreLoggerConfiguration)

        Dim data As DataStoreLoggerConfiguration = New DataStoreLoggerConfiguration()

        If configuration IsNot Nothing Then
            configuration.Invoke(data)
        End If

        Return Function() data

    End Function

End Module

File: CSharp/Core/Serilog.Sinks.LogView.Core/Extensions/DataStoreLoggerSinkExtensions.cs
using Serilog.Configuration;
using LogViewer.Core;

namespace Serilog.Sinks.LogView.Core;

public static class DataStoreLoggerSinkExtensions
{
    public static LoggerConfiguration DataStoreLoggerSink
    (
        this LoggerSinkConfiguration loggerConfiguration,
        Func<ILogDataStore> dataStoreProvider, 
        Action<DataStoreLoggerConfiguration>? configuration = null,
        IFormatProvider formatProvider = null!
    )
        => loggerConfiguration.Sink(new DataStoreLoggerSink(dataStoreProvider, GetConfig(configuration), formatProvider));

    private static Func<DataStoreLoggerConfiguration> GetConfig(Action<DataStoreLoggerConfiguration>? configuration)
    {
        // convert from Action to Func delegate to pass data
        DataStoreLoggerConfiguration data = new();
        configuration?.Invoke(data);
        return () => data;
    }
}